clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -47,32 +185,166 @@ public boolean isValid() {
+        if (!isCompletingTransaction) {
+            TransactionalRequestResult result = new TransactionalRequestResult();
+            FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+
+            isCompletingTransaction = true;
+            if (!newPartitionsToBeAddedToTransaction.isEmpty()) {
+                pendingTransactionalRequests.add(addPartitionsToTransactionRequest(false));
+            }
+            pendingTransactionalRequests.add(endTxnRequest(isCommit, false, result));
+            return resultFuture;
         }
-        return pidAndEpoch;
+        return null;
+    }
+
+    public synchronized FutureTransactionalResult sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                         String consumerGroupId) {
+        TransactionalRequestResult result = new TransactionalRequestResult();
+        FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+        pendingTransactionalRequests.add(addOffsetsToTxnRequest(offsets, consumerGroupId, false, result));
+        return resultFuture;
+    }
+
+    public boolean isInTransaction() {
+        return isTransactional() && isInTransaction;
+    }
+
+
+    public synchronized void maybeAddPartitionToTransaction(TopicPartition topicPartition) {
+        if (partitionsInTransaction.contains(topicPartition))
+            return;
+        newPartitionsToBeAddedToTransaction.add(topicPartition);
+    }
+
+    public void needsRetry(TransactionalRequest request) {
+        request.setRetry();
+        pendingTransactionalRequests.add(request);
+    }
+
+    public void didNotSend(TransactionalRequest request) {
+        pendingTransactionalRequests.add(request);
+    }
+
+    public Node coordinator(FindCoordinatorRequest.CoordinatorType type) {
+        switch (type) {
+            case GROUP:
+                return consumerGroupCoordinator;
+            case TRANSACTION:
+                return transactionCoordinator;
+            default:
+                return null;
+        }
+    }
+
+    public void needsCoordinator(TransactionalRequest request) {
+        needsCoordinator(request.coordinatorType, request.coordinatorKey);
+    }
+
+    private void needsCoordinator(FindCoordinatorRequest.CoordinatorType type, String coordinatorKey) {
+        switch (type) {
+            case GROUP:
+                consumerGroupCoordinator = null;
+                break;
+            case TRANSACTION:
+                transactionCoordinator = null;
+        }
+        pendingTransactionalRequests.add(findCoordinatorRequest(type, coordinatorKey, false));
+    }
+
+
+    public void setInFlightRequestCorrelationId(int correlationId) {
+        inFlightRequestCorrelationId = correlationId;
+    }
+
+    public void resetInFlightRequestCorrelationId() {
+        inFlightRequestCorrelationId = Integer.MIN_VALUE;
```

Comment:
We should probably just use `-1` for the unset value? Or `null`. The latter is a little safer, but either option seems easier to recognise than a really large negative number.